core/src/main/scala/caliban/parsing/VariablesUpdater.scala

```diff
@@ -7,39 +7,72 @@ import caliban.parsing.adt.Type.{ ListType, NamedType }
 import caliban.parsing.adt._
 import caliban.schema.RootType
 import caliban.{ GraphQLRequest, InputValue, Value }
+import zio._
+import caliban.CalibanError.ValidationError
 
 object VariablesUpdater {
-  def updateVariables(
+  def prepare(
     req: GraphQLRequest,
     doc: Document,
     rootType: RootType
-  ): GraphQLRequest = {
+  ): IO[ValidationError, GraphQLRequest] = {
     val variableDefinitions = doc.operationDefinitions.flatMap(_.variableDefinitions)
-    val updated             = req.variables.getOrElse(Map.empty).map { case (key, value) =>
-      val v =
-        variableDefinitions
-          .find(_.name == key)
-          .map { definition =>
-            rewriteValues(value, definition.variableType, rootType)
-          }
-          .getOrElse(value)
+    var variables           = req.variables.getOrElse(Map.empty)
+
+    IO.foldLeft(variableDefinitions)(Map.empty[String, InputValue]) { case (coercedValues, definition) =>
+      IO.fromEither(isInputType(definition.variableType, rootType))
+        .mapError(e => ValidationError(s"Type of variable '${definition.name}' $e", "")) *> {
+        val value =
+          variables
+            .get(definition.name)
+            .map(rewriteValues(_, definition.variableType, rootType))
+            .orElse(definition.defaultValue.map(IO.succeed(_)))
+
+        (value, definition.variableType.nonNull) match {
+          case (None, true)  => IO.fail(ValidationError("null for non-null", ""))
+          case (None, false) => IO.succeed(coercedValues)
+          case (Some(v), _)  =>
+            v.map(value => coercedValues + (definition.name -> value))
+        }
+      }
+    }.map(coercedValues => req.copy(variables = Some(coercedValues)))
+  }
 
-      key -> v
+  // https://spec.graphql.org/June2018/#IsInputType()
+  private def isInputType(t: Type, rootType: RootType): Either[String, Unit] =
+    t match {
+      case NamedType(name, nonNull)  =>
+        rootType.types
+          .get(name)
+          .map { t =>
+            isInputType(t).left
+              .map(_ => s"is not a valid input type.")
+          }
+          .getOrElse({ Left(s"is not a valid input type.") })
+      case ListType(ofType, nonNull) =>
+        isInputType(ofType, rootType).left.map(_ => s"is not a valid input type.")
     }
 
-    req.copy(variables = Some(updated))
+  private def isInputType(t: __Type): Either[__Type, Unit] = {
```

Comment:
This is very similar to what's in `Validator.scala`, but couldn't find a great way to have a single one